libtcmu_log.c

```diff
@@ -365,3 +392,146 @@ static struct log_buf *tcmu_log_initialize(void)
 	pthread_mutex_unlock(&g_mutex);
 	return logbuf;
 }
+
+static int append_output(log_output_func f, log_close_func c, void *data,
+				int priority, int dest, const char *name)
+	
+{
+	int ret = -1;
+	char *ndup = NULL;
+	struct log_output output;
+	if (f == NULL)
+	    return -1;
+
+	if (dest == TCMU_LOG_TO_FILE) {
+		if (name == NULL)
+			return -1;
+		ndup = strdup(name);
+		if (ndup == NULL)
+	    		return -1;
+	}
+
+	output.f = f; 
+	output.c = c; 
+	output.data = data;
+	output.priority = priority;
+	output.dest = dest;
+	output.name = ndup;
+
+	darray_append(logbuf->outputs, output);
+
+	return ret;
+}
+
+static int output_to_syslog(int priority, const char *timestamp, 
+				const char *str, void *data)
+{
+	syslog(priority, "%s", str);
+	return strlen(str);
+}
+
+
+static void close_syslog(void *data)
+{
+	closelog();
+}
+
+static void close_fd(void *data)
+{
+	int fd = (intptr_t) data;
+	close(fd);	
+}
+
+static int add_output_to_syslog(int priority, const char *ident)
+{
+	open_syslog(ident, 0 ,0);
+	if (append_output(output_to_syslog, close_syslog, NULL,
+				priority, TCMU_LOG_TO_SYSLOG, ident) < 0) {
+		closelog();
+		return -1;
+	}
+	return 0;
+}
+
+static int output_to_fd(int priority, const char *timestamp,
+				const char *str,void *data)
+{
+	int fd = (intptr_t) data;
+	char *buf, *msg; 
+	int count, ret, written, r, pid = 0;
+
+	if (fd < 0) 
+		return -1;
+
+	pid = getpid();
+	if (pid <= 0)
+		return -1;
+
+	/* 
+	 * format: timestamp pid [loglevel] msg
+	 */
+	ret = asprintf(&msg, "%s %d [%s] %s", timestamp, pid, loglevel_string(priority), str);	
+	if (ret < 0)
+		return -1;
+
+	buf = msg;
+
+	/* safe write */
+	count = strlen(buf);	
+	while (count > 0) {
+		r = write(fd, buf, count);
+		if (r < 0 && errno == EINTR)
```

Comment:
What was the reason for handling EINTR? Were you hitting this in normal use or just handling it to be thorough?